### PR TITLE
Adding conversion from Seurat 3 to Monocle 2

### PR DESCRIPTION
**Pull request by miretchin**
Given the popularity of these two packages, I wrote conversion functions to move the data types between them.

I've added conversion from Seurat object to CellDataSet (CDS) Monocle 2 file format, and vice versa. Also, as necessary, I added a `as.CellDataSet` generic in `generic.R` .

I would love your suggestions!

Best,
Michael

### CODE REVIEW

File: R/generics.R [hunk 1]
```diff
@@ -33,8 +33,19 @@ AddMetaData <- function(object, metadata, col.name = NULL) {
...
 
-#' Convert a matrix (or Matrix) to the Graph class.
+#' Convert objects to CellDataSet objects
+#'
+#' @param x An object to convert to class \code{CellDataSet}
+#'
+#' @rdname as.CellDataSet
+#' @export as.CellDataSet
+#'
+as.CellDataSet <- function(x, ...) {
```
Comment (mojaveazure) [inline R/generics.R:44]: Be sure to document all parameters, including the `...`

File: R/objects.R [hunk 2]
```diff
@@ -1486,6 +1486,66 @@ UpdateSeuratObject <- function(object) {
 # Methods for Seurat-defined generics
 #%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
 
+#' @param assay Assay to convert
+#'
+#' @rdname as.CellDataSet
+#' @export
+#' @method as.CellDataSet Seurat
+#'
+as.CellDataSet.Seurat <- function(x, assay = NULL, dim.res = NULL, ...) {
```
Comment (mojaveazure) [inline R/objects.R]: Be sure to document all parameters

File: R/objects.R [hunk 3]
```diff
@@ -1486,6 +1486,66 @@ UpdateSeuratObject <- function(object) {
...
+  if (!PackageCheck('monocle', error = FALSE)) {
+    stop("Please install monocle from Bioconductor before converting to a CellDataSet object")
+  }
+  assay <- assay %||% DefaultAssay(object = x)
+  # make variables, then run `newCellDataSet`
+  # create cellData counts
+  counts <- GetAssayData(object = x, assay = assay, slot = "counts")
+  # create phenoData
+  pd <- tryCatch(
+    pd <- new("AnnotatedDataFrame", data = x@meta.data),
```
Comment (mojaveazure) [inline R/objects.R]: Cell-level metadata can be accessed by `x[[]]`

File: R/objects.R [hunk 4]
```diff
@@ -1486,6 +1486,66 @@ UpdateSeuratObject <- function(object) {
...
+  }
+  # no meaningful way to get the lower expression limit
+  cds <- monocle::newCellDataSet(
+    cellData = counts,
+    phenoData = pd,
+    featureData = fd,
+    expressionFamily = expressionFamily
+  )
+  # adding dimensionality reduction data to the CDS
+  for (dr in names(x = x@reductions)) {
```
Comment (mojaveazure) [inline R/objects.R]: Names of dimensional reductions can be found with `FilterObjects`

File: R/objects.R [hunk 5]
```diff
@@ -1486,6 +1486,66 @@ UpdateSeuratObject <- function(object) {
...
+    cds@auxClusteringData[[dr]] <- x@reductions[[dr]]@cell.embeddings
```
Comment (mojaveazure) [inline R/objects.R]: Cell embedding matrices can be accessed using `Embeddings`

File: R/objects.R [hunk 6]
```diff
@@ -1525,6 +1585,66 @@ as.Graph.matrix <- function(x, ...) {
   return(as.Graph.Matrix(x = as(object = x, Class = 'Matrix')))
 }
 
+#' @param counts name of the CellDataSet assay to store as \code{counts}
+#' @param data name of the CellDataSet assay to slot as \code{data}
```
Comment (mojaveazure) [inline R/objects.R]: Parameter `data` is neither defined nor used in the function. Will it be used or can we remove it?

File: R/objects.R [hunk 7]
```diff
@@ -1525,6 +1585,66 @@ as.Graph.matrix <- function(x, ...) {
...
+#' @rdname as.Seurat
+#' @export
+#' @method as.Seurat CellDataSet
+#'
+as.Seurat.CellDataSet <- function(x, counts = "counts", ...) {
+  if (!PackageCheck('monocle', error = FALSE)) {
+    stop("Please install monocle from Bioconductor before converting to a CellDataSet object")
+  }
+  counts <- tryCatch(
+    expr = exprs(x),
```
Comment (mojaveazure) [inline R/objects.R]: `exprs` was not imported. Remember to import all functions used with Roxygen comments in the documentation

```R
#' @importFrom <packageName> exprs
```

File: R/objects.R [hunk 8]
```diff
@@ -1525,6 +1585,66 @@ as.Graph.matrix <- function(x, ...) {
...
+    error = function(e) {
+      stop("No data in provided assay - ", counts)
+    }
+  )
+  meta.data <- as.data.frame(x = pData(x))
```
Comment (mojaveazure) [inline R/objects.R]: `pData` was not imported

File: R/objects.R [hunk 9]
```diff
@@ -1525,6 +1585,66 @@ as.Graph.matrix <- function(x, ...) {
...
+      if (dim(x = embeddings)[2] == length(x = cell.names)) {
+        # Flip so columns are coordinates
+        embeddings <- t(x = embeddings)
+        # Make sure the cell/row names match.
+        # rownames(x = embeddings) <- cell.names
+        rownames(x = embeddings) <- colnames(x = seurat.object)
+        key <- gsub(
+          pattern = "[[:digit:]]",
+          replacement = "_",
+          x = colnames(x = x@auxClusteringData[[dr]])[1]
```
Comment (mojaveazure) [inline R/objects.R]: Pleased use `slot` or a Monocle-provided accessor function instead of the `@` sigil.

File: R/objects.R [hunk 10]
```diff
@@ -1601,10 +1721,10 @@ as.SingleCellExperiment.Seurat <- function(x, assay = NULL, ...) {
     stop("Please install SingleCellExperiment from Bioconductor before converting to a SingeCellExperiment object")
   }
   assay <- assay %||% DefaultAssay(object = x)
-  sce <- SingleCellExperiment::SingleCellExperiment(assays = list(
-    counts = GetAssayData(object = x, assay = assay, slot = "counts"),
+  cds <- monocle::newCellDataSet(
```
Comment (mojaveazure) [inline R/objects.R]: Is there a reason why we're using `monocle::newCellDataSet` in `as.SingleCellExperiment`?

**Comment by dwucsf** [comment]
I just this development version of Seurat 3 (3.0.1.9020), using:

remove.packages('Seurat')
install.packages('devtools')
devtools::install_github(repo = 'satijalab/seurat', ref = 'develop')
library(Seurat)

I successfully downloaded and installed this version of Seurat 3, but none of the help pages are displaying. I'm getting the following error:

Error in fetch(key) : lazy-load database '/Library/Frameworks/R.framework/Versions/3.5/Resources/library/Seurat/help/Seurat.rdb' is corrupt

Is there a reason for why this is? Thank you all for working on this valuable resource!!

**Comment by dwucsf** [comment]
Also, with this new conversation capability, will the CDS be able to accept both original RNA counts and integrated RNA counts from a Seurat 3 object?

**Comment by mojaveazure** [comment]
@dwucsf,

Please see [here](https://stackoverflow.com/questions/30424608/error-in-fetchkey-lazy-load-database) for corrupt help databases. Monocle v2 is, as far as we can tell, single-assay. As such, you can only convert one assay from your `Seurat` object to a `CellDataSet`

**Comment by Dragonmasterx87** [comment]
Guys, really dumb question, but how is this code deployed? imagine I have a Seurat object generated by v3.0.0, called 'seurat_object', what would the coding be till you hit:

'HSMM <- orderCells(HSMM)
plot_cell_trajectory'

Just want to compare to what I do, cheers!

**Pull request merged**